### ISSUE TITLE
Add .NET support

### ISSUE THREAD

**Issue by benaadams** [pyroscope-io/pyroscope#117]
There is already a cross platform (Windows, macOS, Linux) sampling profiler for .NET as part of the platform which can be invoked with `dotnet trace collect` ([instructions](https://github.com/dotnet/diagnostics/blob/master/documentation/dotnet-trace-instructions.md))

Would be great it this could be adapted or utilised to serve as an agent for pyroscope. 

Have also raised an issue in the .NET diagnostics repo regarding this https://github.com/dotnet/diagnostics/issues/2068

**Comment by juniortads** [comment]
I would like to hear about the next steps on this issue :)

**Comment by petethepig** [comment]
Done some research, here is a rough TODO list:

* [ ] (optional) add windows support, moved this to a separate issue (#166)
* [ ] as for the integration itself, we could add a new `dotnetspy` and copy the architecture after [`ebpfspy` integration we already have](https://github.com/pyroscope-io/pyroscope/tree/main/pkg/agent/ebpfspy). That one starts a script (it's `profile.py` in ebpfspy case, for dotnetspy it will be `dotnet trace collect`), parses the profiling data from `folded` format to internal pyroscope format and the rest is handled by pyroscope agent code.
* [ ] add an example to `examples/dotnet`
* [ ] update docs

**Comment by petethepig** [comment]
So it looks like default dotnet format (`nettrace`) is hard to parse. I think for now we could use `dotnet trace` with `--format speedscope` argument, speedscope format looks much simpler. Plus if we added support for speedscope format it would also make it easier to create other integrations in the future.

**Comment by benaadams** [comment]
Also does `--format Chromium` which is the one used by browser tools

**Comment by kolesnikovae** [comment]
There is one more option to get data from .Net runtime - to use [Diagnostics IPC](https://github.com/dotnet/diagnostics/blob/main/documentation/design-docs/ipc-protocol.md) directly, bypassing `dotnet-trace` tool. This allows us to have more accurate measurements, eliminate some overhead caused by additional serialisation, and be independent of `dotnet-trace` (which is not installed by default). I’ve already implemented a basic IPC client that can access raw [EventPipe](https://github.com/microsoft/perfview/blob/main/src/TraceEvent/EventPipe/EventPipeFormat.md) stream and now I’m working on the decoder.

What do you think guys @benaadams, @petethepig?

**Comment by benaadams** [comment]
> There is one more option to get data from .Net runtime - to use [Diagnostics IPC](https://github.com/dotnet/diagnostics/blob/main/documentation/design-docs/ipc-protocol.md) directly, bypassing `dotnet-trace` tool. This allows us to have more accurate measurements, eliminate some overhead caused by additional serialisation, and be independent of `dotnet-trace` (which is not installed by default). I’ve already implemented a basic IPC client that can access raw [EventPipe](https://github.com/microsoft/perfview/blob/main/src/TraceEvent/EventPipe/EventPipeFormat.md) stream and now I’m working on the decoder.

@tommcdon @josalem would this be a good approach?

**Comment by josalem** [comment]
The CLI tools are (mostly) thin wrappers around the Microsoft.Diagnostics.NetCore.Client NuGet package. That has everything you need to start an EventPipe session (or other diagnostic tasks) without having to spin your own Diagnostics IPC Protocol implementation. As for parsing the event stream, you can use `EventPipeEventSource` from the Microsoft.Diagnostics.TraceEvent package. That has a callback mechanism for inspecting events live or parsing a stream of events that's been written to disk (in nettrace format). The TraceEvent package also has logic for examining stack samples.

Those components are all written in C#, so if you want to write the agent in something other than that, you will have to spin your own implementation for now.

I'd be happy to chat some more about your use case if you have any questions 😄.

**Comment by petethepig** [comment]
@josalem Thanks for the pointers! This is very useful.

@kolesnikovae @benaadams I think we're gonna go with the Diagnostics IPC route for the reasons @kolesnikovae mentioned:
* eliminate some overhead caused by additional serialization
* allows us to have more accurate measurements
* no dependency on `dotnet-trace`
